### PR TITLE
Update interoperability sidechain registration - Closes #7440

### LINKED ISSUE #7440: Update sidechain Registration command on mainchain
### Description

Based on [PR#148](https://github.com/LiskHQ/lips/pull/148), update sidechain registration command.

- Update [`sidechainRegParams`](https://github.com/LiskHQ/lips/blob/1dfda3f1781149ee5c10a55c921f6e01e30201eb/proposals/lip-0043.md#parameters)
    - `genesisBlockID`->`chainID`
    - Add `sidechainRegistrationFee`
    - Add min/max lengths according to LIP
- Update [`verify`](https://github.com/LiskHQ/lips/blob/1dfda3f1781149ee5c10a55c921f6e01e30201eb/proposals/lip-0043.md#verification) method
    - `trs.params.name` should be checked for specified 
    - Remove `netID` calculation and check
    - Add `chainID` check
    - Add extra command fee check and balance check
- Update [`execute`](https://github.com/LiskHQ/lips/blob/1dfda3f1781149ee5c10a55c921f6e01e30201eb/proposals/lip-0043.md#execution) method
    - Update the `execute` function as specified in the [LIP0043](https://github.com/LiskHQ/lips/blob/1dfda3f1781149ee5c10a55c921f6e01e30201eb/proposals/lip-0043.md#execution)

### Acceptance Criteria

- All updates mentioned should be present
- All the test cases should pass based on the changes

### Additional Information

[PR#148](https://github.com/LiskHQ/lips/pull/148) is still open so make sure to double check the recent changes.

### PR DESCRIPTION
**Pull request by mitsuaki-u**
### What was the problem?

This PR resolves #7440 

### How was it solved?

- 🌱 Added constructor to interop module
- 🌱 Add `token module` dependency to interop module
- 🌱 Added 2 events to interop module
- 🌱 Updated `verifyCertificateSignature` 
- 🌱 Updated sidechain reg schemas
- 🌱 Updated verify and execute logic in `sidechain registration command`
- 🌱  Updated `registration cross chain command` for `networkIdentifier` -> `chainID`
- 🌱  Updated `initGenesisStateutil` for `networkIdentifier` -> `chainID` 
- ✅ Updated all tests in interop module to reflect networkID -> chainID

### How was it tested?
- Updated tests

### CODE REVIEW

File: framework/src/modules/interoperability/constants.ts [hunk 1]
```diff
@@ -40,6 +41,8 @@ export const MAX_UINT32 = 4294967295;
 export const MAX_UINT64 = BigInt('18446744073709551615'); // BigInt((2 ** 64) - 1) - 1
 export const THRESHOLD_MAINCHAIN = 68;
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/constants.ts:42]: That is unrelated to this PR. Same for below comment

File: framework/src/modules/interoperability/events/ccm_processed.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+			fieldNumber: 2,
+		},
+	},
+};
+
+export class CcmProcessedEvent extends BaseEvent<CcmProcessedEventData> {
+	public schema = ccmProcessedEventSchema;
+
+	public log(
+		ctx: EventQueuer,
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/events/ccm_processed.ts:42]: I am following the convention in the `BaseEvent` class, which uses the name `ctx`. I will keep it and you can create another issue to change it as it is unrelated to this PR.

File: framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts [hunk 3]
```diff
@@ -69,7 +79,9 @@ export class SidechainRegistrationCommand extends BaseInteroperabilityCommand {
 		if (!isValidName(name)) {
 			return {
 				status: VerifyStatus.FAIL,
-				error: new Error(`Sidechain name is in an unsupported format: ${name}`),
+				error: new Error(
+					`Invalid name property. It should contain only characters from the set [a-z0-9!@$&_.].`,
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts:83]: The error message comes from the LIP PR: https://github.com/LiskHQ/lips/pull/148. I will keep it as is.

File: framework/src/modules/interoperability/constants.ts [hunk 4]
```diff
@@ -40,6 +41,8 @@ export const MAX_UINT32 = 4294967295;
...
 export const MESSAGE_TAG_CERTIFICATE = 'LSK_CE_';
+export const MIN_CHAIN_NAME_LENGTH = 1;
+export const MAX_CHAIN_NAME_LENGTH = 32;
 
 // Chain status
 export const CHAIN_REGISTERED = 0;
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/constants.ts:48]: These constants are specified in the LIP: https://github.com/LiskHQ/lips/pull/148. I will keep it as is.

File: framework/src/modules/interoperability/constants.ts [hunk 5]
```diff
@@ -56,10 +59,16 @@ export const CCM_STATUS_CROSS_CHAIN_COMMAND_NOT_SUPPORTED = 2;
 export const CCM_STATUS_CHANNEL_UNAVAILABLE = 3;
 export const CCM_STATUS_RECOVERED = 4;
 export const MIN_RETURN_FEE = BigInt(1000);
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/constants.ts:61]: Unrelated to this PR

File: framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts [hunk 6]
```diff
@@ -146,107 +164,122 @@ export class SidechainRegistrationCommand extends BaseInteroperabilityCommand {
...
-		await registeredNetworkIDsSubstore.set(
-			context,
-			networkID,
-			{ id: chainIDBuffer },
-			// Note: Uses chainIDSchema
-		);
+		// Burn the registration fee
+		await this._tokenAPI.burn(apiContext, senderAddress, TOKEN_ID_LSK, REGISTRATION_FEE);
+
+		// Emit chain account updated event.
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts:247]: I am following the convention in the BaseEvent class. Events are added to the queue and then emitted.

File: framework/src/modules/interoperability/mainchain/module.ts [hunk 7]
```diff
@@ -41,6 +44,22 @@ export class MainchainInteroperabilityModule extends BaseInteroperabilityModule
 		this.interoperableCCAPIs,
 	);
 
+	public commands = [
+		new SidechainRegistrationCommand(
+			this.stores,
+			this.events,
+			this.interoperableCCAPIs,
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/mainchain/module.ts:51]: As discussed, we will follow what is in the feature branch until it is updated, not development branch.

**Comment by mitsuaki-u** [comment]
@sitetester @ishantiw As discussed and agreed upon in our meeting and noting here:  1. Comments unrelated to the code changed in the PR or unrelated to the PR issue should be addressed in a separately. 2. Constant names will be kept as stated in LIP: https://github.com/LiskHQ/lips/pull/148 (but error messages can be updated).  I will indicate with a "👎" and resolve above related comments and address the rest.

File: framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts:83]: I will leave as is as it is easier to see immediately which characters are allowed or not this way.

File: framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts [hunk 8]
```diff
@@ -80,20 +92,26 @@ export class SidechainRegistrationCommand extends BaseInteroperabilityCommand {
...
 				status: VerifyStatus.FAIL,
-				error: new Error('Name substore must not have an entry for the store key name'),
+				error: new Error('Name already registered.'),
 			};
 		}
 
-		const networkID = utils.hash(Buffer.concat([genesisBlockID, transaction.senderAddress]));
+		// 	The chainID has to be unique with respect to the set of already registered sidechains.
+		const chainDataSubstore = this.stores.get(ChainAccountStore);
+		const chainDataExists = await chainDataSubstore.has(context, chainID);
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts:101]: IMO `chainDataExists` is more clear and easier to read

File: framework/src/modules/interoperability/constants.ts [hunk 9]
```diff
@@ -56,10 +59,16 @@ export const CCM_STATUS_CROSS_CHAIN_COMMAND_NOT_SUPPORTED = 2;
...
+export const CROSS_CHAIN_COMMAND_REGISTRATION = 'crossChainCommandRegistration';
+export const CCM_SENT_STATUS_SUCCESS = 0;
 
 // Commands
 export const COMMAND_NAME_SIDECHAIN_REG = 'sidechainRegistration';
```
Comment (Incede) [inline framework/src/modules/interoperability/constants.ts:66]:
```suggestion
export const COMMAND_SIDECHAIN_REG = 'sidechainRegistration';
```

**Review by Incede — CHANGES_REQUESTED** [review]

File: framework/src/modules/interoperability/constants.ts [hunk 10]
```diff
@@ -19,6 +19,7 @@ export const MODULE_NAME_INTEROPERABILITY = 'interoperability';
 // General constants
 export const MAINCHAIN_ID = 1;
 export const MAINCHAIN_ID_BUFFER = utils.intToBuffer(MAINCHAIN_ID, 4);
+export const TOKEN_ID_LSK = Buffer.from([0, 0, 0, 1, 0, 0, 0, 0]);
```
Comment (ishantiw) [inline framework/src/modules/interoperability/constants.ts:22]: I think it should be `000000000000` after chainID but maybe we should wait for Token LIP update on this one. Please make a comment as `TODO: To be updated after token module update `

File: framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (shuse2) [reply framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts:249]: I think @sitetester means comment and the function name is not matching?

[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts:249]: I think comment is clear and conveys that an event will be emitted. Also the comment comes from the LIP. Will keep as is if it's ok.

File: framework/src/modules/interoperability/events/ccm_processed.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+		sendingChainID: Buffer,
+		receivingChainID: Buffer,
+		ccmID: Buffer,
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/events/ccm_processed.ts:45]: Yes according to the LIP it should be included

[the remark this answers is not in the document]
Comment (shuse2) [reply framework/src/modules/interoperability/events/ccm_processed.ts:45]: yes, but i meant here
```
public log(
		ctx: EventQueuer,
		sendingChainID: Buffer,
		receivingChainID: Buffer,
//		ccmID: Buffer,  <= this ccmID can be taken from data
		data: CcmProcessedEventData,
	): void {
```

File: framework/src/modules/interoperability/stores/chain_account.ts [hunk 12]
```diff
@@ -29,23 +29,18 @@ export interface LastCertificateJSON {
...
 	name: string;
-	networkID: Buffer;
 	lastCertificate: LastCertificate;
 	status: number;
 }
 
 const chainAccountJSONSchema = {
 	type: 'object',
-	required: ['name', 'networkID', 'lastCertificate', 'status'],
+	required: ['name', 'lastCertificate', 'status'],
```
Comment (ishantiw) [inline framework/src/modules/interoperability/stores/chain_account.ts:38]: update the fieldNumber for 'lastCertificate', 'status'

**Review by ishantiw — CHANGES_REQUESTED** [review]
Overall LGTM, couple of minor comments

**Review by ishantiw — APPROVED** [review]
🚀

File: framework/test/unit/modules/interoperability/mainchain/commands/sidechain_registration.spec.ts [hunk 13]
```diff
@@ -160,22 +212,25 @@ describe('Sidechain registration command', () => {
...
-			expect(result.error?.message).toInclude(
-				'Name substore must not have an entry for the store key name',
-			);
+			expect(result.error?.message).toInclude('Name already registered.');
 		});
 
-		it('should return error if store key networkID already exists in networkID store', async () => {
-			await networkIDSubstore.set(createStoreGetter(stateStore), networkID, {
-				id: utils.intToBuffer(0, 4),
+		it('should return error if store key chainID already exists in chainID store', async () => {
```
Comment (shuse2) [inline framework/test/unit/modules/interoperability/mainchain/commands/sidechain_registration.spec.ts:218]: i think cases for 
* ` trs.params.chainID[0] != CHAIN_ID_MAINCHAIN[0]:`
* sidechainRegistrationFee !== REGISTRATION_FEE
* vailableBalance < REGISTRATION_FEE
are missing?

File: framework/test/unit/modules/interoperability/mainchain/commands/sidechain_registration.spec.ts [hunk 14]
```diff
@@ -550,24 +505,53 @@ describe('Sidechain registration command', () => {
...
+				REGISTRATION_FEE,
+			);
+		});
+
+		it(`should emit ${EVENT_NAME_CHAIN_ACCOUNT_UPDATED} event`, async () => {
+			// Act
+			await sidechainRegistrationCommand.execute(context);
+
+			// Assert
+			expect(chainAccountUpdatedEvent.log).toHaveBeenCalled();
```
Comment (shuse2) [inline framework/test/unit/modules/interoperability/mainchain/commands/sidechain_registration.spec.ts:531]: these should also check the input as well? same as other event expectations

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts [hunk 15]
```diff
@@ -146,107 +164,124 @@ export class SidechainRegistrationCommand extends BaseInteroperabilityCommand {
...
+			localID: utils.intToBuffer(0, 4),
+		};
 		const channelSubstore = this.stores.get(ChannelDataStore);
-		await channelSubstore.set(context, chainIDBuffer, {
+		await channelSubstore.set(context, chainID, {
 			inbox: { root: EMPTY_HASH, appendPath: [], size: 0 },
 			outbox: { root: EMPTY_HASH, appendPath: [], size: 0 },
 			partnerChainOutboxRoot: EMPTY_HASH,
-			messageFeeTokenID: { chainID: utils.intToBuffer(1, 4), localID: utils.intToBuffer(0, 4) },
+			messageFeeTokenID,
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts:228]: `TOKEN_ID_LSK` has a different type from what is required in `messageFeeTokenID`

File: framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts [hunk 16]
```diff
@@ -146,107 +164,124 @@ export class SidechainRegistrationCommand extends BaseInteroperabilityCommand {
...
+		});
+
+		const ownChainAccountSubstore = this.stores.get(OwnChainAccountStore);
+		const ownChainAccount = await ownChainAccountSubstore.get(methodContext, EMPTY_BYTES);
+
+		const ccm = {
+			nonce: ownChainAccount.nonce,
+			module: MODULE_NAME_INTEROPERABILITY,
+			crossChainCommand: CROSS_CHAIN_COMMAND_REGISTRATION,
+			sendingChainID: ownChainAccount.id,
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/mainchain/commands/sidechain_registration.ts:266]: no `id` is correct

File: framework/src/modules/interoperability/base_interoperability_store.ts [hunk 17]
```diff
@@ -147,11 +147,14 @@ export abstract class BaseInteroperabilityStore {
...
+		const chainAccounts = await chainSubstore.iterate(this.context, {
+			gte: startChainID,
+			lte: endBuf,
+		});
 
 		return Promise.all(
-			chainIDs.map(async chainID => {
-				const chainIDBuffer = utils.intToBuffer(chainID.key.readUInt32BE(0), 4);
+			chainAccounts.map(async chainAccount => {
+				const chainIDBuffer = utils.intToBuffer(chainAccount.key.readUInt32BE(0), 4);
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/interoperability/base_interoperability_store.ts:157]: This code is from a different PR. I merely updated the variable names to be more clear but I can update it if the code in the previous PR was invalid.

**Review by shuse2 — APPROVED** [review]
🚀

**Pull request merged**